src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -264,6 +264,95 @@ struct pj {
     }
 };
 
+template<json_schema_dialect Dialect>
+json::SchemaDocument const& get_metaschema() {
+    static auto const meteschema_doc = [] {
+        auto metaschema_str = [] {
+            switch (Dialect) {
+            case json_schema_dialect::draft4:
+                return json_draft_4_metaschema;
+            }
+        }();
+
+        auto metaschema_json = json::Document{};
+        metaschema_json.Parse(metaschema_str.data(), metaschema_str.size());
+        vassert(
+          !metaschema_json.HasParseError(), "Malformed metaschema document");
+
+        return json::SchemaDocument{metaschema_json};
+    }();
+
+    return meteschema_doc;
+}
+
+result<void> validate_json_schema(
+  json_schema_dialect dialect, json::Document const& schema) {
+    // validation pre-step: get metaschema for json draft
+    auto const& metaschema_doc = [=]() -> json::SchemaDocument const& {
+        using enum json_schema_dialect;
+        switch (dialect) {
+        case json_schema_dialect::draft4:
+            return get_metaschema<draft4>();
+        }
+    }();
+
+    // validation of schema: validate it against metaschema
+    auto validator = json::SchemaValidator{metaschema_doc};
+
+    if (!schema.Accept(validator)) {
+        // schema it's not a valid json schema for Dialect, according to
+        // metaschema
+
+        auto error_loc_metaschema = json::StringBuffer{};
+        auto error_loc_schema = json::StringBuffer{};
+        validator.GetInvalidSchemaPointer().StringifyUriFragment(
+          error_loc_metaschema);
+        validator.GetInvalidDocumentPointer().StringifyUriFragment(
+          error_loc_schema);
+        auto invalid_keyword = validator.GetInvalidSchemaKeyword();
+
+        return error_info{
+          error_code::schema_invalid,
+          fmt::format(
+            "Invalid json schema: '{}', invalid metaschema: '{}', invalid "
```

Comment:
This section seems to be duplicating code from a previous commit.

Reply:
The change from the first to the last commit was large; I tried to produce a cleaner history by adding the new code before removing the old one. That's why some blocks are duplicated in the intermediate stage